File: cpp/rotational-cipher/rotational_cipher.cpp
#include "rotational_cipher.h"
#include <iostream>

namespace rotational_cipher
{
    std::string rotate(std::string input, int rotation)
    {
        std::string characters = "abcdefghijklmnopqrstuvwxyz";
        std::string encoded;
        for (const char &ch : input)
        {
            if (std::isalpha(ch))
            {
                if (std::islower(ch))
                {
                    encoded += characters[(characters.find(ch) + rotation) % 26];
                    continue;
                }
                if (std::isupper(ch))
                {
                    encoded += std::toupper(characters[(characters.find(std::tolower(ch)) + rotation) % 26]);
                    continue;
                }
            }
            encoded += ch; // not alphanum
        }
        return encoded;
    }
} // namespace rotational_cipher

File: go/rotational-cipher/rotational_cipher.go
package rotationalcipher

import (
	"strings"
	"unicode"
)

func RotationalCipher(plain string, shiftKey int) string {
	var output strings.Builder

	for _, v := range plain {
		if unicode.IsPunct(v) || unicode.IsDigit(v) {
			output.WriteString(string(v))
			continue
		}

		if unicode.IsSpace(v) {
			output.WriteString(" ")
			continue
		}

		if unicode.IsUpper(v) {
			output.WriteString(strings.ToUpper(shift(v, shiftKey)))
			continue
		}

		if unicode.IsLower(v) {
			output.WriteString(shift(v, shiftKey))
			continue
		}
	}

	return output.String()
}

func shift(char rune, shiftKey int) string {
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	start := strings.Index(alphabet, strings.ToLower(string(char)))
	return string(alphabet[(start+shiftKey)%len(alphabet)])
}
